src/libdredd/src/mutation_replace_binary_operator.cc

```diff
@@ -202,31 +268,94 @@ std::string MutationReplaceBinaryOperator::GetFunctionName(
 
+  // In the case that we can optimise out some binary expressions, it is
+  // important to change the name of the mutator function to avoid clashes
+  // with other versions that apply to the same operator and types but cannot
+  // be optimised.
+  if (!binary_operator_.isAssignmentOp()) {
+    clang::Expr::EvalResult eval_result;
+    bool rhs_is_int =
+        binary_operator_.getRHS()->EvaluateAsInt(eval_result, ast_context);
+    if (rhs_is_int && llvm::APSInt::isSameValue(eval_result.Val.getInt(),
+                                                llvm::APSInt::get(0))) {
+      result += "_rhs_zero";
+    } else if (rhs_is_int && llvm::APSInt::isSameValue(eval_result.Val.getInt(),
+                                                       llvm::APSInt::get(1))) {
+      result += "_rhs_one";
+    } else if (rhs_is_int && llvm::APSInt::isSameValue(eval_result.Val.getInt(),
+                                                       llvm::APSInt::get(-1))) {
+      result += "_rhs_minus_one";
+    }
+
+    bool lhs_is_int =
+        binary_operator_.getLHS()->EvaluateAsInt(eval_result, ast_context);
+    if (lhs_is_int && llvm::APSInt::isSameValue(eval_result.Val.getInt(),
+                                                llvm::APSInt::get(0))) {
+      result += "_lhs_zero";
+    } else if (lhs_is_int && llvm::APSInt::isSameValue(eval_result.Val.getInt(),
+                                                       llvm::APSInt::get(1))) {
+      result += "_lhs_one";
+    } else if (lhs_is_int && llvm::APSInt::isSameValue(eval_result.Val.getInt(),
+                                                       llvm::APSInt::get(-1))) {
+      result += "_lhs_minus_one";
+    }
+  }
+
   return result;
 }
 
 void MutationReplaceBinaryOperator::GenerateArgumentReplacement(
     const std::string& arg1_evaluated, const std::string& arg2_evaluated,
+    clang::ASTContext& ast_context, bool optimise_mutations,
     std::stringstream& new_function, int& mutant_offset) const {
   if (!binary_operator_.isAssignmentOp()) {
     // LHS
-    new_function << "  if (__dredd_enabled_mutation(local_mutation_id + "
-                 << mutant_offset << ")) return " << arg1_evaluated << ";\n";
-    mutant_offset++;
+    // These cases are equivalent to constant replacement with the respective
+    // constants
+    clang::Expr::EvalResult lhs_eval_result;
+    bool lhs_is_int =
+        binary_operator_.getLHS()->EvaluateAsInt(lhs_eval_result, ast_context);
+    if (!optimise_mutations ||
+        !(lhs_is_int && (llvm::APSInt::isSameValue(lhs_eval_result.Val.getInt(),
+                                                   llvm::APSInt::get(0)) ||
+                         llvm::APSInt::isSameValue(lhs_eval_result.Val.getInt(),
+                                                   llvm::APSInt::get(1)) ||
+                         llvm::APSInt::isSameValue(lhs_eval_result.Val.getInt(),
+                                                   llvm::APSInt::get(-1))))) {
```

Comment:
I am worried that we might lose something when the LHS is *unsigned*, but has a constant value that would evaluate to  -1 if interpreted as signed. In that case we wouldn't do the -1 replacement in constant replacement mode. Can you dig into this a bit?

Reply:
After experimenting, this fortunately doesn't appear to be an issue. Even when an unsigned integer operand could be evaluated as -1, we still replace the expression with both operands. I have added a test case to check this.